Updating the encoding of the Issuance Authorization keys and signatures

vivek-arte opened the pull request:
This PR makes the updates to the encoding of the issuance validating key and the issuance authorization signature, as done in the specification in zcash/zips#1042, along with the further updates in zcash/zips#1048 and zcash/zips#1053. 

The test vectors are updated in QED-it/zcash-test-vectors#31.

vivek-arte committed:
adding IssuanceAuthorizationSignature struct wrapping around schnorr::Signature

vivek-arte committed:
updating IssuanceAuthorizationSignature struct to wrap [u8; 64] instead of schnorr::Signature

vivek-arte committed:
updating IssuanceValidatingKey struct to wrap [u8; 32] instead of schnorr::VerifyingKey

vivek-arte committed:
updating IssuanceAuthorizingKey struct to wrap [u8; 32] instead of NonZeroScalar

vivek-arte committed:
pulling out the issuance authorization key material into its own module to separate out the issuance authorization material

vivek-arte committed:
shifting IssuanceAuthorizationSignature to issuance_auth module

vivek-arte committed:
adding Scheme ID enum to all the issuance auth signature structs (currently mostly hardcoded to ZSASchnorr though)

vivek-arte committed:
adding trait to work with scheme enum to the structs, and generalizing

vivek-arte committed:
adding encode and decode to ik impl, adding TryFrom impl for the scheme ID enum, updating asset base derivation and test vectors

vivek-arte committed:
initial functional generic implementation

vivek-arte commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,422 @@
+    /// The length in bytes of an issuance validating key.
+    const IK_LENGTH: usize;
+    /// The length in bytes of an issuance authorization signature.
+    const SIG_LENGTH: usize;
```
Removed these associated constants, and also removed the checks in the `from_bytes` functions.

vivek-arte commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,422 @@
+    ///
+    /// Returns `None` if the bytes do not correspond to a valid issuance authorizing key.
+    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
+        (bytes.len() == S::ISK_LENGTH)
```
Removed these associated constants, and also removed the checks in the `from_bytes` functions.

vivek-arte commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,422 @@
+
+    /// Deserialize the issuance authorization signature from its canonical byte representation.
+    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
+        (bytes.len() == S::SIG_LENGTH)
```
Removed these associated constants, and also removed the checks in the `from_bytes` functions.

vivek-arte commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,436 @@
+///
+/// [issueauthsig]: https://zips.z.cash/zip-0227#issuance-authorization-signature-scheme
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct IssuanceAuthorizationSignature {
```
Added the generics, also removed the `scheme` field, since it is not necessary with generics.

vivek-arte commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,422 @@
+    ///
+    /// [issuancekeycomponents]: https://zips.z.cash/zip-0227#derivation-of-issuance-validating-key
+    pub fn encode(&self) -> Vec<u8> {
+        let mut encoded = vec![S::ALGORITHM_BYTE];
```
Done.

vivek-arte commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,422 @@
+    ///
+    /// [issueauthsig]: https://zips.z.cash/zip-0227#issuance-authorization-signing-and-validation
+    pub fn encode(&self) -> Vec<u8> {
+        let mut encoded = vec![S::ALGORITHM_BYTE];
```
Done.

vivek-arte commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,422 @@
+    /// Real issuance keys should be derived according to [ZIP 32].
+    ///
+    /// [ZIP 32]: https://zips.z.cash/zip-0032
+    pub fn random(rng: &mut impl CryptoRngCore) -> Self {
```
Added `random_isk` to the trait. Do you recommend also adding the `#[cfg(test)]`?

vivek-arte committed:
responding to review comments

ConstanceBeguier commented on src/note/asset_base.rs:
```diff
@@ -189,6 +201,12 @@ pub mod testing {
+#[cfg(test)]
+mod tests {
+    use crate::issuance_auth::ZSASchnorrSigScheme;
+    use crate::{issuance_auth::IssuanceValidatingKey, note::AssetBase};
```
You could put all crate imports together

vivek-arte commented on src/circuit/note_commit.rs:
```diff
@@ -2648,6 +2648,7 @@ mod tests {
     };
     use pasta_curves::{arithmetic::CurveAffine, pallas, EpAffine};
 
+    use crate::issuance_auth::ZSASchnorrSigScheme;
```
Done.

vivek-arte commented on src/circuit/value_commit_orchard.rs:
```diff
@@ -157,6 +157,7 @@ mod tests {
     use pasta_curves::pallas;
 
     use crate::circuit::value_commit_orchard::ZsaValueCommitParams;
+    use crate::issuance_auth::ZSASchnorrSigScheme;
```
Done, with a little more such optimization

vivek-arte commented on src/issuance.rs:
```diff
@@ -1847,24 +1887,25 @@ mod tests {
 #[cfg(any(test, feature = "test-dependencies"))]
 #[cfg_attr(docsrs, doc(cfg(feature = "test-dependencies")))]
 pub mod testing {
+    use crate::issuance::IssuanceAuthorizationSignature;
```
Done

vivek-arte commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,417 @@
+// Preserve '::' which specifies the EXTERNAL 'zip32' crate
+#[rustfmt::skip]
+pub use ::zip32::{AccountId, ChildIndex, DiversifierIndex, Scope, hardened_only};
+use crate::issuance::Error;
```
Done. Though actually, is there a benefit to having a separate Error for `issuance_auth`? I considered it, but then this is all mostly used in `issuance`, so I think using `issuance::Error` for everything is reasonable.

ConstanceBeguier replied:
I think also it is reasonable

vivek-arte committed:
optimizing import lines across the crate

vivek-arte commented on src/note/asset_base.rs:
```diff
@@ -122,9 +134,9 @@ impl AssetBase {
-    /// This is only used in tests.
-    pub(crate) fn random(rng: &mut impl CryptoRngCore) -> Self {
-        let isk = IssuanceAuthorizingKey::random(rng);
+    /// This is only used in tests. //TODO: VA: This will only work for ZSA Schnorr signatures. Generalization would need random isk generation for other schemes.
```
Done.

vivek-arte commented on src/note/asset_base.rs:
```diff
@@ -46,15 +48,29 @@ pub const ZSA_ASSET_DIGEST_PERSONALIZATION: &[u8; 16] = b"ZSA-Asset-Digest";
+    asset_desc_hash: &[u8; 32],
+) -> Vec<u8> {
+    let mut asset_id = vec![version];
+    asset_id.extend_from_slice(&ik.encode()); // TODO: VA: Do we need to panic here if decoding is not 33 bytes?
```
Agreed, actually we are taking the encoding from a `IssuanceValidatingKey`, so we expect that to be implemented correctly, no need to check it here.

vivek-arte commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,417 @@
+
+    #[test]
+    fn issuance_authorizing_key_from_bytes_fail_on_zero() {
+        // isk must not be the zero scalar for the ZSA Schnorr scheme. TODO: VA: I think this needs to be ZSA Schnorr specific - can't claim this for all schemes.
```
Done.

vivek-arte commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,417 @@
+    fn try_sign(isk: &Self::IskType, msg: &[u8; 32]) -> Result<Self::IssueAuthSigType, Error> {
+        schnorr::SigningKey::from(*isk)
+            .sign_prehash(msg)
+            .map_err(|_| issuance::Error::IssueBundleInvalidSignature)
```
Yes, makes sense, my bad. Fixed.

vivek-arte commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,417 @@
+
+/// A key used to validate issuance authorization signatures, denoted by `ik`.
+///
+/// Defined in [ZIP 227: Issuance of Zcash Shielded Assets § Issuance Key Generation][IssuanceZSA].
```
Done.

vivek-arte commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,417 @@
+    ) -> Result<(), issuance::Error>;
+}
+
+/// An issuance authorization key, from is used to sign the issuance authorization signatures.
```
Done.

vivek-arte committed:
applying suggestions from review - mostly comment fixes

vivek-arte commented on src/issuance.rs:
```diff
@@ -1613,8 +1652,9 @@ mod tests {
-        let incorrect_isk = IssuanceAuthorizingKey::random(&mut rng);
-        let incorrect_ik: IssuanceValidatingKey = (&incorrect_isk).into();
+        let incorrect_isk: IssuanceAuthorizingKey<ZSASchnorrSigScheme> =
+            IssuanceAuthorizingKey::random(&mut rng);
```
Done

vivek-arte committed:
further work on review comments

vivek-arte committed:
removing IssueAuthSigScheme generic from IssueBundle and Signed authorization

vivek-arte committed:
renaming ZSASchnorrSigScheme to ZSASchnorr

vivek-arte committed:
updating the issue bundle commitment to also take the algorithm byte prefix

vivek-arte commented on src/issuance.rs:
```diff
@@ -164,7 +167,10 @@ impl IssueAction {
     ///    Pallas curve.
     /// * `IssueActionWithoutNoteNotFinalized`: The `IssueAction` contains no notes and is not finalized.
-    fn verify(&self, ik: &IssuanceValidatingKey) -> Result<(AssetBase, NoteValue), Error> {
+    fn verify<S: IssuanceAuthSigScheme>(
```
Done

vivek-arte commented on src/issuance.rs:
```diff
@@ -238,33 +244,51 @@ pub struct Prepared {
-pub struct Signed {
-    signature: schnorr::Signature,
+#[derive(PartialEq, Eq)]
+pub struct Signed<S: IssuanceAuthSigScheme> {
```
Done

vivek-arte committed:
Apply suggestions from code review

Co-authored-by: Constance Beguier <constance@qed-it.com>

vivek-arte committed:
additions based on review comments

vivek-arte commented on src/keys.rs:
```diff
@@ -1207,13 +1043,13 @@ mod tests {
-            let ik: IssuanceValidatingKey = (&isk).into();
-            assert_eq!(ik.to_bytes(), tv.ik);
+            // let ik: IssuanceValidatingKey = (&isk).into();
+            // assert_eq!(ik.to_bytes(), tv.ik);
```
Added them back

ConstanceBeguier commented:
I’m surprised that only asset_base.rs was modified in the test_vectors folder.
I think keys.rs should also be modified.

ConstanceBeguier reviewed (approved):
Approved pending the following comments
- Please add a description to the PR
- Check which files into test_vectors folder should be updated
- Check with Pablo whether the current trait architecture is acceptable (I was not able to simplify it)

vivek-arte committed:
updating keys test_vectors

vivek-arte committed:
Merge branch 'zsa1' into ik_isk_issueauthsig

PaulLaux commented on src/note/asset_base.rs:
```diff
@@ -46,15 +47,31 @@ pub const ZSA_ASSET_DIGEST_PERSONALIZATION: &[u8; 16] = b"ZSA-Asset-Digest";
+/// Encoding the Asset Identifier, as defined in [ZIP 227][assetidentifier].
+///
+/// [assetidentifier]: https://zips.z.cash/zip-0227.html#specification-asset-identifier-asset-digest-and-asset-base
+pub fn encode_asset_id<S: IssuanceAuthSigScheme>(
```
Avoid generalizing this file. Lock everything to `<ZSASchnorr>` similar to `IssueBundle()`. (all functions in `asset_base.rs`)

vivek-arte replied:
Done.

PaulLaux commented on src/issuance.rs:
```diff
@@ -238,21 +245,22 @@ pub struct Prepared {
+#[derive(Clone, Debug, PartialEq, Eq)]
 pub struct Signed {
-    signature: schnorr::Signature,
+    signature: IssuanceAuthorizationSignature<ZSASchnorr>,
```
rename the type to `IssueAuthSig<_>`

vivek-arte replied:
Done

PaulLaux commented on src/issuance.rs:
```diff
@@ -164,7 +168,10 @@ impl IssueAction {
-    fn verify(&self, ik: &IssuanceValidatingKey) -> Result<(AssetBase, NoteValue), Error> {
+    fn verify(
+        &self,
+        ik: &IssuanceValidatingKey<ZSASchnorr>,
```
rename type to `IssueValidatingKey<_>`

vivek-arte replied:
Done

PaulLaux commented on src/issuance.rs:
```diff
@@ -539,8 +547,11 @@ fn create_reference_note(asset: AssetBase, mut rng: impl RngCore) -> Note {
-        let expected_ik: IssuanceValidatingKey = isk.into();
+    pub fn sign(
+        self,
+        isk: &IssuanceAuthorizingKey<ZSASchnorr>,
```
rename type to `IssueAuthKey<_>`

vivek-arte replied:
Done

PaulLaux commented on src/issuance.rs:
```diff
@@ -704,7 +715,13 @@ pub enum Error {
+
     /// Verification errors:
+    /// Invalid issuance validating key.
+    InvalidIssuanceValidatingKey,
```
Please make it shorter, similar to the new names of the types mentioned above. Also, ensure consistency of the other errors (IssueBundleInvalidSignature is inconsistent).

ConstanceBeguier replied:
Done

PaulLaux commented on src/issuance.rs:
```diff
@@ -834,8 +855,8 @@ mod tests {
-        let isk = IssuanceAuthorizingKey::random(&mut rng);
-        let ik: IssuanceValidatingKey = (&isk).into();
+        let isk = IssuanceAuthorizingKey::<ZSASchnorr>::random(&mut rng);
+        let ik = (&isk).into();
```
let's be explisit here: `let ik = IssuanceValidatingKey::from(&isk);
`

vivek-arte replied:
Done

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,416 @@
+//! Issuance logic for Zcash Shielded Assets (ZSAs).
+//!
+//! This module provides the structures and methods necessary for handling issuance authorization
+//! signatures and the issuance keys.
```
be more detailed + include code example
```rust
...
let isk = IssuanceAuthorizingKey::<ZSASchnorr>::random(&mut rng);
let ik = IssuanceValidatingKey::from(&isk);
... sign
...  verify
```

ConstanceBeguier replied:
Done

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,416 @@
+        coin_type: u32,
+        account: u32,
+    ) -> Result<Self, zip32::Error> {
+        // Call zip32 logic
```
spec says account MUST be zero so
```rust
if account != 0 {
        return Err(zip32::Error::InvalidDerivationPath);
    }
```
or similar

ConstanceBeguier replied:
Done

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,416 @@
+
+    fn try_sign(isk: &Self::IskType, msg: &[u8; 32]) -> Result<Self::IssueAuthSigType, Error> {
+        schnorr::SigningKey::from(*isk)
+            .sign_prehash(msg)
```
we must use the 0 aux randomness. can we move to 
` sign_raw(&self, msg: &[u8], aux_rand: &[u8; 32])` and force `aux_rand=0` ?

ConstanceBeguier replied:
Done

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,416 @@
+            .map_err(|_| issuance::Error::InvalidIssuanceAuthorizingKey)
+    }
+
+    fn verify(ik: &Self::IkType, msg: &[u8], sig: &Self::IssueAuthSigType) -> Result<(), Error> {
```
~~lets fix the size of the `msg` to `&[u8; 32]` and call it `verify_sig()`~~ (edit: not needed)

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,416 @@
+    fn try_sign(isk: &Self::IskType, msg: &[u8; 32]) -> Result<Self::IssueAuthSigType, Error> {
+        schnorr::SigningKey::from(*isk)
+            .sign_prehash(msg)
+            .map_err(|_| issuance::Error::InvalidIssuanceAuthorizingKey)
```
there could be multiple reasons.
`IssueBundleInvalidSignature` seems to fit better

ConstanceBeguier replied:
Done

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,416 @@
+impl<S: IssuanceAuthSigScheme> Debug for IssuanceAuthorizingKey<S> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> core::fmt::Result {
+        f.debug_tuple("IssuanceAuthorizingKey")
+            .field(&self.to_bytes())
```
we probably want to trancate and present only the last 4 bytes on debug

ConstanceBeguier replied:
IssueValidatingKey now displays only the last four bytes.
IssueAuthKey displays the last four bytes of its derived IssueValidatingKey.
I prefer not to reveal any bytes from IssueAuthKey itself.

ConstanceBeguier replied:
Done

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,416 @@
+
+    fn verify(ik: &Self::IkType, msg: &[u8], sig: &Self::IssueAuthSigType) -> Result<(), Error> {
+        ik.verify_prehash(msg, sig)
+            .map_err(|_| issuance::Error::IssueBundleInvalidSignature)
```
`issuance::` redundent (all)

vivek-arte replied:
Done

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,416 @@
+    /// Serialization for the issuance authorization signature.
+    fn sig_to_bytes(sig: &Self::IssueAuthSigType) -> Vec<u8>;
+    /// Deserialization for the issuance authorization signature.
+    fn sig_from_bytes(bytes: &[u8]) -> Option<Self::IssueAuthSigType>;
```
This return `None` on fail while all our functions return `Result<>`. 
For consistency, convert eveithing possibble to `Result`. Here included.

ConstanceBeguier replied:
All other ..._from_bytes functions in Orchard return  either an Option or a CtOption.
So, I prefer to keep the Option.

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,417 @@
+    use proptest::prelude::*;
+
+    prop_compose! {
+        /// Generate a uniformly distributed Orchard issuance authorizing key. TODO: VA: Can we generalize prop_compose with generics?
```
ok, remove comment

vivek-arte replied:
Done

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,416 @@
+    fn ik_from_bytes(bytes: &[u8]) -> Option<Self::IkType>;
+    /// Serialization for the issuance authorization signature.
+    fn sig_to_bytes(sig: &Self::IssueAuthSigType) -> Vec<u8>;
+    /// Deserialization for the issuance authorization signature.
```
As discussed, we dont need all these functions public. Remove them from the trait and provide only the needed external api

See related: https://github.com/QED-it/orchard/pull/182#discussion_r2306972273

ConstanceBeguier replied:
Done

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,416 @@
+#[rustfmt::skip]
+pub use ::zip32::{AccountId, ChildIndex, DiversifierIndex, Scope, hardened_only};
+
+const ZIP32_PURPOSE: u32 = 32;
```
what is this (unused anyway) ?

vivek-arte replied:
Removed. Was copied from `issuance.rs` and not removed inadvertently.

PaulLaux commented on src/note/asset_base.rs:
```diff
@@ -76,19 +93,15 @@ impl AssetBase {
     #[allow(non_snake_case)]
-    pub fn derive(ik: &IssuanceValidatingKey, asset_desc_hash: &[u8; 32]) -> Self {
-        let version_byte = [0x00];
+    pub fn derive<S: IssuanceAuthSigScheme>(
```
same

vivek-arte replied:
Done

PaulLaux commented on src/note/asset_base.rs:
```diff
@@ -123,8 +136,8 @@ impl AssetBase {
     /// This is only used in tests.
-    pub(crate) fn random(rng: &mut impl CryptoRngCore) -> Self {
-        let isk = IssuanceAuthorizingKey::random(rng);
+    pub(crate) fn random<S: IssuanceAuthSigScheme>(rng: &mut impl CryptoRngCore) -> Self {
```
same

vivek-arte replied:
Done

PaulLaux reviewed (commented):
Overall good work. Added comments to reduce footprint and fixes.
Also,
update `pub const ZIP32_ORCHARD_ISSUANCE_PERSONALIZATION: &[u8; 16] = b"ZIP32ZSAIssue_V1` to `ZcashSA_Issue_V1`as required by zip (please verify)

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,416 @@
+    }
+}
+
+impl<S: IssuanceAuthSigScheme> IssuanceValidatingKey<S> {
```
This level of abstruction is not needed for all functions:

to help removing `ik_from_isk()` and all other function from the public trait, implement only specific versions and not general:

```rust
impl IssuanceAuthorizingKey<ZSASchnorr> {
...
    pub fn to_bytes(&self) -> [u8; 32] { self.0.to_bytes() } //specific implementation without the need for S::ik_from_isk()
}

impl IssuanceValidatingKey<ZSASchnorr> {..}

impl IssuanceAuthorizationSignature<ZSASchnorr> {..}
```

and keep the once that use the public functions generic:

```rust
impl<S: IssuanceAuthSigScheme> IssuanceAuthorizingKey<S> {
    pub fn to_validating_key(&self) -> IssuanceValidatingKey<S> {
        IssuanceValidatingKey(S::ik_from_isk(&self.0))
    }
    pub fn try_sign(&self, msg32: &[u8; 32])
        -> Result<IssuanceAuthorizationSignature<S>, issuance::Error>
    {
        S::try_sign(&self.0, msg32).map(IssuanceAuthorizationSignature)
    }
}
impl<S: IssuanceAuthSigScheme> IssuanceValidatingKey<S> {
    pub fn verify(&self, msg32: &[u8; 32], sig: &IssuanceAuthorizationSignature<S>)
        -> Result<(), issuance::Error>
    {
        S::verify(&self.0, msg32, &sig.0)
    }
}
```

ConstanceBeguier replied:
Done

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,416 @@
+        let isk_bytes = isk.to_bytes();
+        let isk_roundtrip = IssuanceAuthorizingKey::<ZSASchnorr>::from_bytes(&isk_bytes).unwrap();
+        assert_eq!(isk_bytes, isk_roundtrip.to_bytes());
+    }
```
Please add round trip for `ik` and `sig`. Also please add a test for fail in signature verification.

ConstanceBeguier replied:
Done

vivek-arte committed:
removing generics from src/note/asset_base.rs functions

vivek-arte committed:
removed unused constant

vivek-arte committed:
refactoring to shorten issuance authorization struct names

vivek-arte committed:
removing unnecessary issuance:: path prefix from Error calls, and making explicit the IssueValidatingKet::from call

vivek-arte committed:
remove discussed TODO comment

ConstanceBeguier committed:
Update ZcashSA_Issue_V1 to ZcashSA_Issue_V1

ConstanceBeguier committed:
Some renamings

IssueBundleInvalidSignature -> InvalidIssueBundleSig
InvalidIssuanceValidatingKey -> InvalidIssueValidatingKey
InvalidIssuanceAuthorizingKey -> InvalidIssueAuthKey

ConstanceBeguier committed:
Add zip32::Error::NonZeroAccount

ConstanceBeguier committed:
Update returned error for try_sign function

ConstanceBeguier committed:
Rename IssueAuthSigScheme to IssueAuthSigScheme

ConstanceBeguier committed:
Update issuance_auth documentation

ConstanceBeguier committed:
Replace (&isk).into() with IssueValidatingKey::from(&isk)

ConstanceBeguier committed:
Simplify IssueAuthSigScheme trait

ConstanceBeguier committed:
Update debug for issuance validating/authorizing keys

ConstanceBeguier committed:
Add try_sing_with_aux and verify_with_aux

ConstanceBeguier committed:
Add tests

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,443 @@
+        }
+
+        schnorr::SigningKey::from(*isk)
+            .sign_prehash(msg)
```
the intent was to replace `schnorr::SigningKey::from(*isk).sign_prehash(msg)`
with `.sign_raw(&self, msg: &[u8], aux_rand: &[u8; 32])` not change out external API

PaulLaux replied:
(just provide 0 explicitly to `schnorr::SigningKey.sign_raw(&self, msg: &[u8], aux_rand: &[u8; 32])`

ConstanceBeguier replied:
Done

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,443 @@
+    ///
+    /// [issuancekeycomponents]: https://zips.z.cash/zip-0227#derivation-of-issuance-validating-key
+    pub fn encode(&self) -> Vec<u8> {
+        let ik_bytes = self.to_bytes();
+        let mut encoded = Vec::with_capacity(1 + ik_bytes.len());
+        encoded.push(ZSASchnorr::ALGORITHM_BYTE);
+        encoded.extend(ik_bytes);
+        encoded
```
same:
```rust
    let mut encoded = vec![ZSASchnorr::ALGORITHM_BYTE];
    encoded.extend(self.to_bytes());
    encoded
```
(all)

ConstanceBeguier replied:
I prefer to use with_capacity because it preallocates the exact size avoiding reallocations.
In addition extend_from_slice copies bytes in one shot which is faster than extend.

PaulLaux replied:
If so, use size of `ZSASchnorr::ALGORITHM_BYTE` + `ik_bytes.len()`, not `+1`

PaulLaux replied:
so this is optimal
```rust
    #[inline]
    pub(crate) fn encode2(&self) -> Vec<u8> {
        let sig = self.to_bytes();
        let exp_capacity = sig.len() + std::mem::size_of_val(&ZSASchnorr::ALGORITHM_BYTE);
        let mut out = Vec::with_capacity(exp_capacity);
        out.push(ZSASchnorr::ALGORITHM_BYTE);
        out.extend_from_slice(&sig);
        out
    }
```
but honstly
```rust
pub(crate) fn encode2(&self) -> Vec<u8> {
    let mut out = vec![ZSASchnorr::ALGORITHM_BYTE];
    out.extend_from_slice(&self.to_bytes());
    out
}
```
you choose:)
(anything just not `+1`)

ConstanceBeguier replied:
Done

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,443 @@
+    /// Decodes an issuance validating key from the byte representation defined in [ZIP 227][issuancekeycomponents].
+    ///
+    /// [issuancekeycomponents]: https://zips.z.cash/zip-0227#derivation-of-issuance-validating-key
+    pub fn decode(bytes: &[u8]) -> Result<Self, Error> {
+        let (&algorithm_byte, key_bytes) = bytes
+            .split_first()
+            .ok_or(Error::InvalidIssueValidatingKey)?;
+
+        if algorithm_byte != ZSASchnorr::ALGORITHM_BYTE {
+            return Err(Error::InvalidIssueValidatingKey);
+        }
+
+        Self::from_bytes(key_bytes).ok_or(Error::InvalidIssueValidatingKey)
+    }
```
consider (optional)
```rust
    pub fn decode(bytes: &[u8]) -> Result<Self, Error> {
        if let Some((&algorithm_byte, key_bytes)) = bytes.split_first() {
            if algorithm_byte == ZSASchnorr::ALGORITHM_BYTE {
                return Self::from_bytes(key_bytes).ok_or(Error::InvalidIssueValidatingKey);
            }
        }
        Err(Error::InvalidIssueValidatingKey)
    }
```

PaulLaux replied:
x2

ConstanceBeguier replied:
Done

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,443 @@
+        let ik_bytes = self.to_bytes();
+        let last4 = &ik_bytes[(ik_bytes.len() - 4)..];
+
+        write!(
+            f,
+            "IssueValidatingKey {{ last4: 0x{:02x}{:02x}{:02x}{:02x} }}",
+            last4.first().copied().unwrap_or(0),
+            last4.get(1).copied().unwrap_or(0),
+            last4.get(2).copied().unwrap_or(0),
+            last4.get(3).copied().unwrap_or(0),
+        )
```
```rust
write!(
            f,
            "IssueValidatingKey {{ last4: 0x{:02x}{:02x}{:02x}{:02x} }}",
            last4[0], last4[1], last4[2], last4[3]
        )
```
no need to check due to line 281

ConstanceBeguier replied:
Done

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,443 @@
+    }
+}
+
+impl Debug for IssueValidatingKey<ZSASchnorr> {
```
move both Debug to the botom, just before tests

ConstanceBeguier replied:
Done

PaulLaux reviewed (commented):
Let's remove
```rust
fn try_sign_with_aux(
        isk: &Self::IskType,
        msg: &[u8; 32],
        aux_data: &[u8; 32],
    ) -> Result<Self::IssueAuthSigType, Error>;

    /// Verifies a signature over a message and auxiliary data using the issuance validating key.
    fn verify_with_aux(
        ik: &Self::IkType,
        msg: &[u8],
        aux_data: &[u8; 32],
        signature: &Self::IssueAuthSigType,
    ) -> Result<(), Error>;
```
from the trait. I don't see a reason to keep it since we are providing this functions directly from the keys. this will simplify more unless I miss somthing.
\+ fix the aux_random misunderstanding.

ConstanceBeguier committed:
Revert "Add try_sing_with_aux and verify_with_aux"

This reverts commit d61ed08f58b1f0c6c1889e3c55859537848c6cd0.

ConstanceBeguier committed:
Use sign_raw

ConstanceBeguier committed:
Update decode

ConstanceBeguier committed:
Simplify IssueValidatingKey Debug

ConstanceBeguier committed:
Move Debug impl just before tests

ConstanceBeguier commented:
> Let's remove
> ```rust
> fn try_sign_with_aux(...)
> fn verify_with_aux(...)
> ```
> from the trait. I don't see a reason to keep it since we are providing this functions directly from the keys. this will simplify more unless I miss somthing.
> + fix the aux_random misunderstanding

In the trait, I replaced try_with_with_aux and verify_with_aux with try_sign and verify.
I prefer to keep try_sign and verify in the trait to emphasize that any new scheme must implement them.
I fixed the aux_data misunderstanding.

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,383 @@
+}
+
+impl IssueAuthSig<ZSASchnorr> {
+    fn to_bytes(&self) -> Vec<u8> {
+        self.0.to_bytes().to_vec()
+    }
+
+    pub(crate) fn from_bytes(bytes: &[u8]) -> Option<Self> {
+        schnorr::Signature::try_from(bytes).ok().map(Self)
+    }
```
I think we should remove `to_bytes()` and `from_bytes()` and keep just `encode()` and `decode()` (potentialy `encode_bytes()` and `decode_bytes()`).

I don't want two options to go to bytes. What do you think?
(x3)

ConstanceBeguier replied:
I agree with you that we will have some confusion if we keep both the to/from_bytes and encode/decode functions.
For IssueAuthSig and IssueValidatingKey, I kept only the encode/decode functions because in the specs, ik and sig are always used with the ALGORITHM_BYTE.
For IssueAuthKey, I kept only the to/from_bytes functions because it is never used with ALGORITHM_BYTE in the spec.

ConstanceBeguier replied:
We have to update orchard_zsa_issuance_auth_sig.rs test vector file to add ALGORITHM_BYTE 0x00 at the beginning of ik and sig.

ConstanceBeguier replied:
Updated orchard_zsa_issuance_auth_sig test vectors

ConstanceBeguier committed:
Update encode functions with size_of_val

ConstanceBeguier committed:
Remove redundant from/to_bytes functions

ConstanceBeguier committed:
Fix no_std issues

ConstanceBeguier committed:
Fix no_std issues v2

ConstanceBeguier committed:
Update issuance_auth_sig test vectors

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -205,29 +199,25 @@ impl IssueValidatingKey<ZSASchnorr> {
-        let mut encoded = Vec::with_capacity(1 + sig_bytes.len());
+        let sig_bytes = self.0.to_bytes().to_vec();
+        let mut encoded = Vec::with_capacity(
+            core::mem::size_of_val(&ZSASchnorr::ALGORITHM_BYTE) + sig_bytes.len(),
```
nice, but import it so we are left with
`( size_of_val(&ZSASchnorr::ALGORITHM_BYTE) + ik_bytes.len() )`
(x2)

ConstanceBeguier replied:
Done

PaulLaux reviewed (approved):
Approved with minor comments

PaulLaux commented on src/issuance_auth.rs:
```diff
@@ -0,0 +1,375 @@
+    }
+
+    fn verify(ik: &Self::IkType, msg: &[u8], sig: &Self::IssueAuthSigType) -> Result<(), Error> {
+        ik.verify_prehash(msg, sig)
```
We are using `verify_prehash()`, is this the best fit? 
The lib has also ` fn verify_digest(()` and `fn verify()`

If we stick to `verify_prehash`, our API +Docs should reflect it: limit the parameters to the correct size/type

ConstanceBeguier replied:
I will keep `verify_prehash` because we will verify a `SIGHASH` that is already a finalized digest.
`verify_digest` does not fit because it takes as input a digest that is not finalized
`verify` does not fit because it hashes the message before verifying it

ConstanceBeguier replied:
I will fix the size of the msg to the size of SIGHASH (32 bytes)

ConstanceBeguier replied:
Done

ConstanceBeguier committed:
Import core::meme::size_of_val

ConstanceBeguier committed:
Fix size of message to verify
